### PR TITLE
Add capacity reservation to spec

### PR DESCRIPTION
**Pull request by Skarlso**
### Description

Closes https://github.com/weaveworks/eksctl/issues/5258

Note, that this adds a bit more than just the ID to be more flexible and preempt future requests of "just add the ARN"... :)

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by Skarlso** [comment]
Anyone knows how to update the docs for the config reference? 
I thought `build-pages` would do the trick, but it didn't.

**Comment by Himangini** [comment]
looks like the schema is correctly updated in the docs
Screenshot 2022-07-15 at 10 21 10

**Review by Himangini — DISMISSED** [review]
LGTM 👍🏻

File: pkg/apis/eksctl.io/v1alpha5/types.go [hunk 1]
```diff
@@ -1431,6 +1431,21 @@ type NodeGroupBase struct {
...
+
+// CapacityReservation defines reservation policy for a nodegroup
+type CapacityReservation struct {
+	CapacityReservationPreference *string                    `json:"capacityReservationPreference,omitempty"`
+	CapacityReservationTarget     *CapacityReservationTarget `json:"capacityReservationTarget,omitempty"`
+}
+
+// CapacityReservationTarget defines reservation target details
+type CapacityReservationTarget struct {
+	CapacityReservationID               *string `json:"capacityReservationId,omitempty"`
```
Comment (nikimanoledaki) [inline pkg/apis/eksctl.io/v1alpha5/types.go:1447]:
```suggestion
	CapacityReservationID               *string `json:"capacityReservationID,omitempty"`
```

shouldn't this be uppercase `ID`?

**Review by nikimanoledaki — DISMISSED** [review]
A couple of comments/change requests! :) 
We could also add validation that `CapacityReservationPreference` is set to a valid value, which is either `open` or `none`, and add a note about this in the schema.

File: pkg/cfn/builder/managed_launch_template.go [hunk 2]
```diff
@@ -19,6 +19,23 @@ func (m *ManagedNodeGroupResourceSet) makeLaunchTemplateData(ctx context.Context
...
+				return gfnt.NewString(*value)
+			}
+			return nil
+		}
+		launchTemplateData.CapacityReservationSpecification = &gfnec2.LaunchTemplate_CapacityReservationSpecification{}
+		launchTemplateData.CapacityReservationSpecification.CapacityReservationPreference = valueOrNil(mng.CapacityReservation.CapacityReservationPreference)
+		if mng.CapacityReservation.CapacityReservationTarget != nil {
+			launchTemplateData.CapacityReservationSpecification.CapacityReservationTarget = &gfnec2.LaunchTemplate_CapacityReservationTarget{
+				CapacityReservationId:               valueOrNil(mng.CapacityReservation.CapacityReservationTarget.CapacityReservationID),
+				CapacityReservationResourceGroupArn: valueOrNil(mng.CapacityReservation.CapacityReservationTarget.CapacityReservationResourceGroupARN),
```
Comment (nikimanoledaki) [inline pkg/cfn/builder/managed_launch_template.go:34]: The [CapacityReservationSpecification](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CapacityReservationSpecification.html) docs specify that:
```
You can specify only one parameter at a time. If you specify CapacityReservationPreference and CapacityReservationTarget, the request fails.
```
We should add validation that not both are set at the same time.

File: pkg/apis/eksctl.io/v1alpha5/types.go (hunk 1 above)
Comment (Skarlso) [reply pkg/apis/eksctl.io/v1alpha5/types.go:1447]: Sure. :D Adjust the tests too :)

**Comment by nikimanoledaki** [comment]
I tried creating it manually with `CapacityReservationTarget` and got this error:
```
You cannot specify a value for CapacityReservationId and CapacityReservationResourceGroupArn in the same request. Specify either a CapacityReservationId or a CapacityReservationResourceGroupArn and try again. 
```

I will add more validation to split it further since those 2 fields also cannot be set at the same time.

**Comment by Himangini** [comment]
> I tried creating it manually with just `CapacityReservationTarget` and both `CapacityReservationTargetId` and `CapacityReservationResourceGroupArn`, and got this error:
> 
> ```
> You cannot specify a value for CapacityReservationId and CapacityReservationResourceGroupArn in the same request. Specify either a CapacityReservationId or a CapacityReservationResourceGroupArn and try again. 
> ```
> 
> I will add more validation to split it further since those 2 fields also cannot be set at the same time.

@nikimanoledaki The original ticket was just about adding the support for CapacityReservationID, I think we should just stick to that and not go overboard with other settings. If we get more requests in the future we'll see how to add support for them. Sounds good?

> Note, that this adds a bit more than just the ID to be more flexible and preempt future requests of "just add the ARN"... :)

**Comment by Skarlso** [comment]
>@nikimanoledaki The original ticket was just about adding the support for CapacityReservationID, I think we should just stick to that and not go overboard with other settings. If we get more requests in the future we'll see how to add support for them. Sounds good?

I don't think she added anything new right now, right? She just improved the UX by returning an error if the fields are defined together. ARN was already there. Thanks, Niki, great work! :) Thanks for shepherding my little baby. :D

**Comment by nikimanoledaki** [comment]
@Himangini The change was simple: adding 1 more validation and updating the tests. I just finished it and pushed the changes. 

Regarding scope, the feature request was to add support for Capacity Reservations (Preference, ID, and Resource Group). This is what this PR does.

**Comment by Himangini** [comment]
> @Himangini The change was simple: adding 1 more validation and updating the tests. I just finished it and pushed the changes.

Thats awesome, thanks for that. 

> 
> Regarding scope, the feature request was to add support for Capacity Reservations (Preference, ID, and Resource Group). This is what this PR does.

With my earlier comment, I was saying I don't think we talked about adding support for ARN during planning, I am guessing this PR is adding extra bits as mentioned in the quote below and all I was saying was if it's creating more problems let's skip the ARN support. 

>Note, that this adds a bit more than just the ID to be more flexible and preempt future requests of "just add the ARN"... :)

**Comment by nikimanoledaki** [comment]
@Himangini Thank you for raising the question regarding scope. Thankfully the extra bits only took some validation to make this feature complete and this PR is ready to be reviewed 👍

**Comment by nikimanoledaki** [comment]
Manually tested that it works:
```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: nm-cap-res-target-id-test
  region: eu-north-1

nodeGroups:
  - name: ng-1
    instanceType: t3.small
    desiredCapacity: 1
    availabilityZones: ["eu-north-1a"]
    capacityReservation:
      capacityReservationTarget:
        capacityReservationID: "cr-07497b08a30634133"
```
It used the Capacity Reservation (0/1 available) 🎉 🥳 
Screenshot 2022-07-21 at 14 20 06

**Review by Himangini — APPROVED** [review]
👍🏻

**Review by cPu1 — APPROVED** [review]

**Comment by tanvp112** [comment]
> looks like the schema is correctly updated in the docs Screenshot 2022-07-15 at 10 21 10

Hi @Himangini , couldn't find the said field at the [online page](https://eksctl.io/usage/schema/). Can you provide the link to the screenshot you captured?

**Comment by Skarlso** [comment]
Hello @tanvp112!

Yes, it applies to both, managed and unmanaged.

The website will not have it yet, as this was just released recently in an RC. :) Please wait for a full release for the docs update. :)

**Pull request merged**